burger/toppings/identify.py

```diff
@@ -105,49 +104,50 @@ def check_match(value, match_list):
 def identify(classloader, path, verbose):
     """
     The first pass across the jar will identify all possible classes it
-    can, maping them by the 'type' it implements.
+    can, mapping them by the 'type' it implements.
 
     We have limited information available to us on this pass. We can only
     check for known signatures and predictable constants. In the next pass,
     we'll have the initial mapping from this pass available to us.
     """
     possible_match = None
 
-    for c in classloader.search_constant_pool(path=path, type_=String):
+    class_file = classloader[path]
+
+    string_constants = classloader.search_constant_pool(path=path, type_=String)
+    for c in string_constants:
```

Comment:
`classloader.search_constant_pool` is a performance optimization to avoid having to parse the whole class file for every class in the game.

Using `time python3 munch.py -t identify -o /dev/null 23w45a.jar`, on my main machine the original code takes about 2.5s to execute before and 5s after. On the Raspberry Pi 2 I use to automatically run Burger for new snapshots, it takes about 65s to execute before and 115s to execute after.

I don't think this is an acceptable slowdown; we should only parse the whole class file if needed. This could maybe be done by removing the `type_=String` part and instead checking with `isinstance`; the Gson field check could be done in the same way.

Reply:
Ah, I made the mistake of assuming search_constant_pool returned a list (which is the reason my chatcomponent detection broke). I also didn't realize how slow classloader was, I just made a commit that should result in it having similar performance to how it was before.